File: server/python/src/metaobjects/render/prompt/output_format_renderer.py
"""FR-010 artifact 1 — output-format prompt renderer ("produce your answer like this").

Renders an :class:`OutputFormatSpec` into a prompt fragment that teaches an LLM how
to shape its answer. Three comment-free styles (guide / inline / exampleOnly) × two
formats (json / xml). Guidance is carried in prose / inline placeholders / a filled
skeleton — NEVER in comments (models ignore them).

Cross-port INVARIANT: the rendered text is byte-identical to the Java/C#/Kotlin/TS
reference (``com.metaobjects.render.prompt.OutputFormatRenderer``). Do not change the
verbatim prose, skeleton shapes, or the numeric-vs-quoted decision.
"""
from __future__ import annotations

import re

import metaobjects.render.escapers as escapers
from metaobjects.render.prompt.output_format_spec import OutputFormatSpec
from metaobjects.render.prompt.prompt_field import PromptField
from metaobjects.render.prompt.prompt_overrides import PromptOverrides
from metaobjects.render.prompt.prompt_style import PromptStyle
from metaobjects.render.recover import FieldKind, Format

_NUMERIC_KINDS: frozenset[FieldKind] = frozenset(
    {FieldKind.INT, FieldKind.LONG, FieldKind.DOUBLE, FieldKind.BOOLEAN}
)

# The render engine OWNS format-keyed escaping; the Format enum's UPPER values
# ("JSON"/"XML") map to the lowercase escaper-registry keys.
def _escape_xml(s: str) -> str:
    return escapers.escape(escapers.FORMAT_XML, s)


def _escape_json(s: str) -> str:
    return escapers.escape(escapers.FORMAT_JSON, s)


def render_output_format(spec: OutputFormatSpec, overrides: PromptOverrides) -> str:
    """Render an :class:`OutputFormatSpec` into an output-format prompt fragment.

    The effective style is the override's style if present, otherwise the spec's.
    """
    effective_style = overrides.style if overrides.style is not None else spec.style
    if effective_style is PromptStyle.EXAMPLE_ONLY:
        return _render_example_only(spec, overrides)
    if effective_style is PromptStyle.INLINE:
        return _render_inline(spec, overrides)
    return _render_guide(spec, overrides)


# ---- INLINE ----------------------------------------------------------------


def _render_inline(spec: OutputFormatSpec, overrides: PromptOverrides) -> str:
    if spec.format is Format.XML:
        return _render_xml_inline(spec, overrides)
    return _render_json_inline(spec, overrides)


def _render_xml_inline(spec: OutputFormatSpec, overrides: PromptOverrides) -> str:
    lines = [
        f"  <{f.name}>{_escape_xml(_inline_content(f, overrides))}</{f.name}>\n"
        for f in spec.fields
    ]
    return f"<{spec.root_name}>\n{''.join(lines)}</{spec.root_name}>"


def _render_json_inline(spec: OutputFormatSpec, overrides: PromptOverrides) -> str:
    lines = [
        f'  "{f.name}": "{_escape_json(_inline_content(f, overrides))}"'
        for f in spec.fields
    ]
    # Empty object is `{\n}` (cross-port parity), not `{\n\n}`.
    if not lines:
        return "{\n}"
    return "{\n" + ",\n".join(lines) + "\n}"


def _inline_content(field: PromptField, overrides: PromptOverrides) -> str:
    if field.kind is FieldKind.ENUM and field.enum_values:
        return " | ".join(field.enum_values)
    if field.kind is FieldKind.BOOLEAN:
        return "true | false"
    instruction = _resolve_instruction(field, overrides)
    if instruction is not None:
        return "{" + instruction + "}"
    return "{" + field.name + "}"


def _resolve_instruction(field: PromptField, overrides: PromptOverrides) -> str | None:
    """Effective instruction: override first, then the field default, else None."""
    ov = overrides.instructions.get(field.name)
    if ov is not None:
        return ov
    return field.instruction


# ---- GUIDE -----------------------------------------------------------------


def _render_guide(spec: OutputFormatSpec, overrides: PromptOverrides) -> str:
    sb = "Fill in each field as described below:\n"
    for field in spec.fields:
        req = "required" if field.required else "optional"
        sb += f"- {field.name} ({req})"
        instruction = _resolve_instruction(field, overrides)
        if instruction is not None:
            sb += f": {instruction}"
        sb += "\n"
        if field.kind is FieldKind.ENUM and field.enum_values:
            sb += f"    one of {', '.join(field.enum_values)}\n"
            enum_doc = field.enum_doc
            if enum_doc is not None:
                for val in field.enum_values:
                    doc = enum_doc.get(val)
                    if doc is not None:
                        sb += f"      {val} = {doc}\n"
        eg = _example_value_if_declared(field, overrides)
        if eg is not None:
            sb += f"    e.g. {eg}\n"
    sb += "\nRespond exactly like this:\n"
    sb += _render_example_only(spec, overrides)
    return sb


# ---- EXAMPLE-ONLY (also the skeleton appended by GUIDE) ---------------------


def _render_example_only(spec: OutputFormatSpec, overrides: PromptOverrides) -> str:
    if spec.format is Format.XML:
        return _render_xml_skeleton(spec, overrides)
    return _render_json_skeleton(spec, overrides)


def _render_xml_skeleton(spec: OutputFormatSpec, overrides: PromptOverrides) -> str:
    lines = [
        f"  <{f.name}>{_escape_xml(_example_value(f, overrides))}</{f.name}>\n"
        for f in spec.fields
    ]
    return f"<{spec.root_name}>\n{''.join(lines)}</{spec.root_name}>"


def _render_json_skeleton(spec: OutputFormatSpec, overrides: PromptOverrides) -> str:
    # NOTE: FieldKind.OBJECT / nested fields are not expanded here — they render as a
    # "{fieldName}" placeholder. Nested-object expansion is a bounded deferral
    # (mirrors Java/C#/TS).
    lines = [
        f'  "{f.name}": {_json_skeleton_value(f, overrides)}' for f in spec.fields
    ]
    # Empty object is `{\n}` (cross-port parity), not `{\n\n}`.
    if not lines:
        return "{\n}"
    return "{\n" + ",\n".join(lines) + "\n}"


def _json_skeleton_value(field: PromptField, overrides: PromptOverrides) -> str:
    """The example value as a JSON literal: bare for numeric/boolean, else quoted."""
    value = _example_value(field, overrides)
    if _is_numeric_or_boolean(field.kind, value):
        return value
    return '"' + _escape_json(value) + '"'


def _example_value_if_declared(field: PromptField, overrides: PromptOverrides) -> str | None:
    from_override = overrides.examples.get(field.name)
    if from_override is not None:
        return from_override
    if field.example is not None:
        return field.example
    return None


def _example_value(field: PromptField, overrides: PromptOverrides) -> str:
    from_override = overrides.examples.get(field.name)
    if from_override is not None:
        return from_override
    if field.example is not None:
        return field.example
    if field.kind is FieldKind.ENUM and field.enum_values:
        return field.enum_values[0]
    return "{" + field.name + "}"


# A canonical ASCII numeric literal. Mirrors the recover engine's `_ASCII_NUMERIC`:
# `[0-9]` (not `\d`) keeps it ASCII-only, rejecting Unicode digits, underscore digit
# grouping, and radix prefixes (0x../0b../0o..) so the emitted JSON stays valid and
# cross-port-identical.
_ASCII_NUMERIC = re.compile(r"^[+-]?(?:[0-9]+\.?[0-9]*|\.[0-9]+)(?:[eE][+-]?[0-9]+)?$")


def _is_numeric_or_boolean(kind: FieldKind, value: str) -> bool:
    if kind not in _NUMERIC_KINDS:
        return False
    if value in ("true", "false"):
        return True
    return bool(_ASCII_NUMERIC.match(value.strip()))

File: server/typescript/packages/render/src/prompt/output-format-renderer.ts
// FR-010 artifact 1 — output-format prompt renderer ("produce your answer like this").
//
// Renders an OutputFormatSpec into a prompt fragment that teaches an LLM how to
// shape its answer. Three comment-free styles (guide / inline / exampleOnly) ×
// two formats (json / xml). Guidance is carried in prose / inline placeholders /
// a filled skeleton — NEVER in comments (models ignore them).
//
// Cross-port INVARIANT: the rendered text is byte-identical to the Java/C#/Kotlin
// reference (com.metaobjects.render.prompt.OutputFormatRenderer). Do not change
// the verbatim prose, skeleton shapes, or numeric-vs-quoted decision.

import { ESCAPERS } from "../escapers.js";
import { FieldKind, Format } from "../extract/types.js";
import type { OutputFormatSpec } from "./output-format-spec.js";
import type { PromptField } from "./prompt-field.js";
import type { PromptOverrides } from "./prompt-overrides.js";
import { PromptStyle } from "./prompt-style.js";

const NUMERIC_KINDS: ReadonlySet<FieldKind> = new Set<FieldKind>([
  FieldKind.INT,
  FieldKind.LONG,
  FieldKind.DOUBLE,
  FieldKind.BOOLEAN,
]);

const INDENT = "  ";
const MAX_NEST_DEPTH = 8;

type SkelMode = "example" | "inline";

// The render engine OWNS format-keyed escaping; Format ("JSON"/"XML") maps to the
// lowercase ESCAPERS keys.
const escapeXml = (s: string): string => ESCAPERS.xml(s);
const escapeJson = (s: string): string => ESCAPERS.json(s);

/**
 * Render an {@link OutputFormatSpec} into an output-format prompt fragment. The
 * effective style is the override's style if present, otherwise the spec's.
 */
export function renderOutputFormat(spec: OutputFormatSpec, overrides: PromptOverrides): string {
  const effectiveStyle = overrides.style ?? spec.style;
  switch (effectiveStyle) {
    case PromptStyle.EXAMPLE_ONLY:
      return renderExampleOnly(spec, overrides);
    case PromptStyle.INLINE:
      return renderInline(spec, overrides);
    default:
      return renderGuide(spec, overrides);
  }
}

// ---- INLINE ----------------------------------------------------------------

function renderInline(spec: OutputFormatSpec, overrides: PromptOverrides): string {
  return spec.format === Format.XML
    ? renderXmlSkeleton(spec, overrides, "inline")
    : renderJsonSkeleton(spec, overrides, "inline");
}

function inlineContent(field: PromptField, overrides: PromptOverrides): string {
  if (field.kind === FieldKind.ENUM && field.enumValues != null && field.enumValues.length > 0) {
    return field.enumValues.join(" | ");
  }
  if (field.kind === FieldKind.BOOLEAN) {
    return "true | false";
  }
  const instruction = resolveInstruction(field, overrides);
  return instruction != null ? `{${instruction}}` : `{${field.name}}`;
}

/** Effective instruction: override first, then the field default, else null. */
function resolveInstruction(field: PromptField, overrides: PromptOverrides): string | null {
  const ov = overrides.instructions?.[field.name];
  if (ov != null) return ov;
  return field.instruction;
}

// ---- GUIDE -----------------------------------------------------------------

function renderGuide(spec: OutputFormatSpec, overrides: PromptOverrides): string {
  let sb = "Fill in each field as described below:\n";
  sb += guideFields(spec, overrides, "", new Set<OutputFormatSpec>([spec]), 0);
  sb += "\nRespond exactly like this:\n";
  sb += renderExampleOnly(spec, overrides);
  return sb;
}

function guideFields(
  spec: OutputFormatSpec, overrides: PromptOverrides, prefix: string,
  path: Set<OutputFormatSpec>, depth: number,
): string {
  let sb = "";
  for (const field of spec.fields) {
    const displayName = prefix + field.name;
    sb += guideEntry(field, overrides, displayName);
    if (canExpand(field, path, depth)) {
      const nested = field.nested!;
      const childPrefix = field.array ? `${displayName}[].` : `${displayName}.`;
      path.add(nested);
      sb += guideFields(nested, overrides, childPrefix, path, depth + 1);
      path.delete(nested);
    }
  }
  return sb;
}

function guideEntry(field: PromptField, overrides: PromptOverrides, displayName: string): string {
  const req = field.required ? "required" : "optional";
  let sb = `- ${displayName} (${req})`;
  const instruction = resolveInstruction(field, overrides);
  if (instruction != null) sb += `: ${instruction}`;
  sb += "\n";
  if (field.kind === FieldKind.ENUM && field.enumValues != null && field.enumValues.length > 0) {
    sb += `    one of ${field.enumValues.join(", ")}\n`;
    const enumDoc = field.enumDoc;
    if (enumDoc != null) {
      for (const val of field.enumValues) {
        const doc = enumDoc[val];
        if (doc != null) sb += `      ${val} = ${doc}\n`;
      }
    }
  }
  const eg = exampleValueIfDeclared(field, overrides);
  if (eg != null) sb += `    e.g. ${eg}\n`;
  return sb;
}

// ---- EXAMPLE-ONLY (also the skeleton appended by GUIDE) ---------------------

function renderExampleOnly(spec: OutputFormatSpec, overrides: PromptOverrides): string {
  return spec.format === Format.XML
    ? renderXmlSkeleton(spec, overrides, "example")
    : renderJsonSkeleton(spec, overrides, "example");
}

// ---- JSON skeleton (recursive) ---------------------------------------------

function renderJsonSkeleton(spec: OutputFormatSpec, overrides: PromptOverrides, mode: SkelMode): string {
  return jsonObject(spec, overrides, "", mode, new Set<OutputFormatSpec>([spec]), 0);
}

function jsonObject(
  spec: OutputFormatSpec, overrides: PromptOverrides, braceIndent: string,
  mode: SkelMode, path: Set<OutputFormatSpec>, depth: number,
): string {
  if (spec.fields.length === 0) return `{\n${braceIndent}}`;
  const fieldIndent = braceIndent + INDENT;
  const lines = spec.fields.map(
    (field) => `${fieldIndent}"${field.name}": ${jsonValue(field, overrides, fieldIndent, mode, path, depth)}`,
  );
  return `{\n${lines.join(",\n")}\n${braceIndent}}`;
}

function jsonValue(
  field: PromptField, overrides: PromptOverrides, indent: string,
  mode: SkelMode, path: Set<OutputFormatSpec>, depth: number,
): string {
  if (field.array) return jsonArray(field, overrides, indent, mode, path, depth);
  if (field.kind === FieldKind.OBJECT) return jsonObjectField(field, overrides, indent, mode, path, depth);
  return jsonLeaf(field, overrides, mode);
}

function jsonLeaf(field: PromptField, overrides: PromptOverrides, mode: SkelMode): string {
  if (mode === "inline") return `"${escapeJson(inlineContent(field, overrides))}"`;
  const value = exampleValue(field, overrides);
  return isNumericOrBoolean(field.kind, value) ? value : `"${escapeJson(value)}"`;
}

function canExpand(field: PromptField, path: Set<OutputFormatSpec>, depth: number): boolean {
  return field.kind === FieldKind.OBJECT && field.nested != null
    && depth < MAX_NEST_DEPTH && !path.has(field.nested);
}

function jsonObjectField(
  field: PromptField, overrides: PromptOverrides, indent: string,
  mode: SkelMode, path: Set<OutputFormatSpec>, depth: number,
): string {
  if (!canExpand(field, path, depth)) return jsonLeaf(field, overrides, mode);
  const nested = field.nested!;
  path.add(nested);
  const out = jsonObject(nested, overrides, indent, mode, path, depth + 1);
  path.delete(nested);
  return out;
}

function jsonArray(
  field: PromptField, overrides: PromptOverrides, indent: string,
  mode: SkelMode, path: Set<OutputFormatSpec>, depth: number,
): string {
  const elemIndent = indent + INDENT;
  let elem: string;
  if (canExpand(field, path, depth)) {
    const nested = field.nested!;
    path.add(nested);
    elem = jsonObject(nested, overrides, elemIndent, mode, path, depth + 1);
    path.delete(nested);
  } else {
    elem = jsonLeaf(field, overrides, mode);
  }
  return `[\n${elemIndent}${elem}\n${indent}]`;
}

// ---- XML skeleton (recursive) ----------------------------------------------

function renderXmlSkeleton(spec: OutputFormatSpec, overrides: PromptOverrides, mode: SkelMode): string {
  return `<${spec.rootName}>\n${xmlBody(spec, overrides, INDENT, mode, new Set<OutputFormatSpec>([spec]), 0)}</${spec.rootName}>`;
}

function xmlBody(
  spec: OutputFormatSpec, overrides: PromptOverrides, indent: string,
  mode: SkelMode, path: Set<OutputFormatSpec>, depth: number,
): string {
  return spec.fields.map((field) => xmlField(field, overrides, indent, mode, path, depth)).join("");
}

function xmlField(
  field: PromptField, overrides: PromptOverrides, indent: string,
  mode: SkelMode, path: Set<OutputFormatSpec>, depth: number,
): string {
  if (canExpand(field, path, depth)) {
    const nested = field.nested!;
    path.add(nested);
    const body = xmlBody(nested, overrides, indent + INDENT, mode, path, depth + 1);
    path.delete(nested);
    return `${indent}<${field.name}>\n${body}${indent}</${field.name}>\n`;
  }
  const content = mode === "inline" ? inlineContent(field, overrides) : exampleValue(field, overrides);
  return `${indent}<${field.name}>${escapeXml(content)}</${field.name}>\n`;
}

function exampleValueIfDeclared(field: PromptField, overrides: PromptOverrides): string | null {
  const ov = overrides.examples?.[field.name];
  if (ov != null) return ov;
  if (field.example != null) return field.example;
  return null;
}

function exampleValue(field: PromptField, overrides: PromptOverrides): string {
  const ov = overrides.examples?.[field.name];
  if (ov != null) return ov;
  if (field.example != null) return field.example;
  if (field.kind === FieldKind.ENUM && field.enumValues != null && field.enumValues.length > 0) {
    return field.enumValues[0]!;
  }
  return `{${field.name}}`;
}

function isNumericOrBoolean(kind: FieldKind, value: string): boolean {
  if (!NUMERIC_KINDS.has(kind)) return false;
  if (value === "true" || value === "false") return true;
  // Finite-only: NaN/Infinity fall through to a quoted string so the emitted JSON
  // stays valid. Number("") is 0, so guard the empty/blank case explicitly. Reject
  // JS-only radix literals (0x../0b../0o..) that Number() accepts but Java/C# don't —
  // same guard as the extract engine's parseFiniteNumber (keeps the JSON valid + parity).
  const t = value.trim();
  if (t === "" || /^[+-]?0[xXbBoO]/.test(t)) return false;
  return Number.isFinite(Number(t));
}
